From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Auke Kok <auke-jan.h.kok@intel.com>
Date: Sat, 3 Nov 2018 20:00:42 -0700
Subject: [PATCH] Use OS provided copy of openssl.cnf as fallback.

This will be used only if OPENSSL_CONF is not set and if
/etc/openssl.cnf is not found. If either one of those is the case, they
will be used and create an error if there is something wrong with them.
---
 apps/apps.c                 | 3 +++
 apps/openssl.c              | 7 ++++++-
 crypto/conf/conf_mod.c      | 7 ++++++-
 include/internal/cryptlib.h | 9 +++++++--
 4 files changed, 22 insertions(+), 4 deletions(-)

diff --git a/crypto/conf/conf_mod.c b/crypto/conf/conf_mod.c
index e703d97..7d3c7ff 100644
--- a/crypto/conf/conf_mod.c
+++ b/crypto/conf/conf_mod.c
@@ -9,6 +9,7 @@
 
 #include "internal/cryptlib.h"
 #include <stdio.h>
+#include <unistd.h>
 #include <ctype.h>
 #include <openssl/crypto.h>
 #include "internal/conf.h"
--- openssl-3.0.5/crypto/conf/conf_mod.c~	2022-07-05 08:57:04.000000000 +0000
+++ openssl-3.0.5/crypto/conf/conf_mod.c	2022-09-11 19:36:20.682992501 +0000
@@ -641,8 +641,10 @@
     if (file == NULL)
         return NULL;
     BIO_snprintf(file, size, "%s%s%s", t, sep, OPENSSL_CONF);
-
-    return file;
+    if (access(file, R_OK) == 0)
+	return file;
+    OPENSSL_free(file);
+    return OPENSSL_strdup("/usr/share/defaults/ssl/openssl.cnf");
 }
 
 /*
--- openssl-3.0.5/apps/lib/apps.c~	2022-07-05 08:57:04.000000000 +0000
+++ openssl-3.0.5/apps/lib/apps.c	2022-09-11 19:42:36.761669443 +0000
@@ -403,6 +403,8 @@
     if (config == NULL)
         config = to_free = app_load_config_quiet(default_config_file);
     if (config == NULL)
+	config = to_free = app_load_config_quiet("/usr/share/defaults/ssl/openssl.cnf");
+    if (config == NULL)
         return 1;
 
     if (CONF_modules_load(config, NULL, 0) <= 0) {
--- openssl-3.2.1/include/internal/common.h~	2024-01-30 13:14:56.000000000 +0000
+++ openssl-3.2.1/include/internal/common.h	2024-01-30 17:51:45.772427721 +0000
@@ -81,9 +81,10 @@
 # define OPENSSL_CONF             "openssl.cnf"
 
 # ifndef OPENSSL_SYS_VMS
+#  define CACHE_CERT_AREA         "/var/cache/ca-certs"
 #  define X509_CERT_AREA          OPENSSLDIR
-#  define X509_CERT_DIR           OPENSSLDIR "/certs"
-#  define X509_CERT_FILE          OPENSSLDIR "/cert.pem"
+#  define X509_CERT_DIR           CACHE_CERT_AREA "/certs"
+#  define X509_CERT_FILE          CACHE_CERT_AREA "/compat/ca-roots.pem"
 #  define X509_PRIVATE_DIR        OPENSSLDIR "/private"
 #  define CTLOG_FILE              OPENSSLDIR "/ct_log_list.cnf"
 # else
